### ISSUE TITLE
How to color-code console output

### ISSUE THREAD

**Issue by ethanabrooks** [ray-project/ray#26226]
Hello. When I use `tune.run`, different processes are distinguished by the number corresponding to their pid:
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/10344742/176714779-ae44e55b-3ec1-4bc5-8f32-ddfe2521e589.png">

I would like to implement color output to distinguish between processes similar to docker-compose:
<img width="1792" alt="image" src="https://user-images.githubusercontent.com/10344742/176718781-fcd12716-8537-4e9f-b9af-0465a494cab2.png">

Could I accomplish this my implementing either a custom `Logger` or a custom `Reporter`? I did some digging but was not able to identify the method responsible for printing the light blue `(trainable pid=1988)` output.

Thanks!

**Comment by xwjiang2010** [comment]
@ethanabrooks 
Love that idea! Do you know what are the things you want to color code that will make reading the log much easier for you (I know that Tune log can be pretty verbose sometimes)? - That decides whether it can be achieved by `Logger`/`Reporter`. 
FYI, the method for printing the light blue `(trainable pid=1988)` output is [here](https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/_private/worker.py?L1620). 

Do you want to contribute a PR? We could definitely help with that!

**Comment by ethanabrooks** [comment]
Sure I would be happy to contribute a PR! I think this would not be too hard to implement. I'll let you know if I have questions.

**Comment by ethanabrooks** [comment]
Also, currently the colors adhere to the convention that yellow is for errors/warnings and cyan is for everything else. Not sure if we would want to add something other than color to distinguish errors/warnings, now that color also indicates different tasks.

**Comment by xwjiang2010** [comment]
@ethanabrooks thanks so much! I get what you are trying to achieve now. I do want to check with Core code owners to make sure this does not overfit to Tune's case.

@rkooo567 @stephanie-wang Could you chime in?
I am wondering 1. does this make sense to apply generically?
2. if 1 is a no, does it make sense to offer this through some env var just for Tune users?

Let me know what you guys think! Thanks!

**Comment by rkooo567** [comment]
@ethanabrooks can you tell me a bit more detail about how the task_index will be used? Alternatively, we can change the color by modifying the printing side code? https://github.com/ray-project/ray/blob/dcc913073f8b0a54f495d24d586a82492f261478/python/ray/_private/worker.py#L790

**Comment by stephanie-wang** [comment]
I like the idea of color-coding the process tag! Agree with @rkooo567 about checking whether we need to add the new task_index field. I'm a bit unclear on how we'll guarantee that this value will stay stable.

The other thing you could try is making the print code pluggable so that you can specify any behavior you want.

**Comment by ethanabrooks** [comment]
@rkooo567 , I was thinking of something like this:
```python
    def color_for(data: Dict[str, str], line: str) -> str:
        """The color for this log line."""
        if (
            data.get("pid") == "raylet"
            and ray_constants.LOG_PREFIX_INFO_MESSAGE not in line
        ):
            return colorama.Fore.YELLOW
        elif data.get("pid") == "autoscaler":
            if "Error:" in line or "Warning:" in line:
                return colorama.Style.BRIGHT + colorama.Fore.YELLOW
            else:
                return colorama.Style.BRIGHT + colorama.Fore.CYAN
        else:
            colors = [
                colorama.Fore.CYAN,
                ...
            ]
            return colors[data["task_index"]]
```
From here: https://github.com/ray-project/ray/blob/dcc913073f8b0a54f495d24d586a82492f261478/python/ray/_private/worker.py#L1653

**Comment by rkooo567** [comment]
Yes. I think although we have task index, there's the risk of conflicting color. Imagine your node has more than 500 workers for example (are we going to have color code as many as Ray's maximum worker proc?) or two task with the same index in different nodes.

**Comment by ethanabrooks** [comment]
So it sounds like using PID is the way to go?

**Comment by stephanie-wang** [comment]
Yes, let's try PID first!

**Comment by rkooo567** [comment]
Hey @ethanabrooks are you planning to work on this?

**Comment by ethanabrooks** [comment]
Yes absolutely! I'm sorry it's been a busy couple weeks. I actually got hung up trying to build the tests and then other things came up.

**Comment by cszhu** [comment]
Hello! This P2 issue has seen no activity in at least 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
